### PR TITLE
Add persistence of available receipt data for message hub

### PR DESCRIPTION
**Pull request by bemwen**
## Description

The purpose of this PR is to add functionality to store information about charge link receipts in the database, so we can later use it to store data for transmission and fetch it again when the market participant is ready for a bundle.

This PR:

* Extends database scripts with new tables for charge link receipts
* Adds repository and tests to store charge link receipt available data
* Renames RequestTime fields in the available data tables in the DB to RequestDateTime to better fit our normal naming
* Refactors charge and charge link available data to use the new AvailableDataBase abstract class

## References

* #823

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/AvailableChargeData.cs [hunk 1]
```diff
@@ -14,18 +14,21 @@
...
 namespace GreenEnergyHub.Charges.Domain.AvailableChargeData
 {
-    public class AvailableChargeData
+    public class AvailableChargeData : AvailableDataBase
     {
         public AvailableChargeData(
             string recipientId,
             MarketParticipantRole recipientRole,
             BusinessReasonCode businessReasonCode,
+            Instant requestTime,
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/AvailableChargeData.cs:30]:
```suggestion
            Instant requestDateTime,
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/AvailableChargeDataFactory.cs [hunk 2]
```diff
@@ -38,6 +38,8 @@ public class AvailableChargeDataFactory : IAvailableChargeDataFactory
                 recipient.Id,
                 recipient.BusinessProcessRole,
                 chargeCommand.Document.BusinessReasonCode,
+                requestTime,
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/AvailableChargeDataFactory.cs:41]:
```suggestion
                requestDateTime,
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeLinkReceiptData/IAvailableChargeLinkReceiptDataRepository.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace GreenEnergyHub.Charges.Domain.AvailableChargeLinkReceiptData
+{
+    public interface IAvailableChargeLinkReceiptDataRepository
+    {
+        Task StoreAsync(List<AvailableChargeLinkReceiptData> availableChargeLinkReceiptData);
+
+        Task<List<AvailableChargeLinkReceiptData>> GetAvailableChargeLinkReceiptDataAsync(IEnumerable<Guid> dataReferenceId);
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeLinkReceiptData/IAvailableChargeLinkReceiptDataRepository.cs:25]:
```suggestion
        Task<List<AvailableChargeLinkReceiptData>> GetAsync(IEnumerable<Guid> dataReferenceId);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeLinksData/AvailableChargeLinksDataFactory.cs [hunk 4]
```diff
@@ -33,15 +33,15 @@ public class AvailableChargeLinksDataFactory : IAvailableChargeLinksDataFactory
                 recipient.Id,
                 recipient.BusinessProcessRole,
                 chargeLinkCommand.Document.BusinessReasonCode,
+                requestTime,
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeLinksData/AvailableChargeLinksDataFactory.cs:36]:
```suggestion
                requestDateTime,
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableData/AvailableDataBase.cs [hunk 5]
```diff
@@ -39,6 +39,13 @@ public abstract class AvailableDataBase
             AvailableDataReferenceId = availableDataReferenceId;
         }
 
+        public AvailableDataBase(string recipientId)
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableData/AvailableDataBase.cs:42]: Nice that we have an available database 😛

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargesDatabaseContext.cs [hunk 6]
```diff
@@ -52,6 +53,8 @@ public ChargesDatabaseContext(DbContextOptions<ChargesDatabaseContext> options)
 
         public DbSet<AvailableChargeData> AvailableChargeData { get; private set; }
 
+        public DbSet<AvailableChargeLinkReceiptData> AvailableChargeLinkReceiptData { get; private set; }
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargesDatabaseContext.cs:56]: Does this hold both confirmations and rejections?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/EntityConfigurations/AvailableChargeLinkReceiptDataConfiguration.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                throw new ArgumentNullException(nameof(builder));
+
+            builder.ToTable("AvailableChargeLinkReceiptData", "MessageHub");
+            builder.HasKey(x => x.Id);
+            builder.Property(p => p.Id).ValueGeneratedNever();
+            builder.Property(x => x.RecipientId).HasColumnName("RecipientId");
+            builder.Property(x => x.RecipientRole).HasColumnName("RecipientRole")
+                .HasConversion(
+                    toDbValue => (int)toDbValue,
+                    fromDbValue => (MarketParticipantRole)fromDbValue);
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/EntityConfigurations/AvailableChargeLinkReceiptDataConfiguration.cs:37]: Can you make a generic enum extension method for all these enum conversions?
```suggestion
            builder
                .Property(x => x.RecipientRole)
                .HasColumnName("RecipientRole")
                .HasEnumConversion<MarketParticipantRole>();
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/AvailableChargeLinkReceiptDataRepository.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            await _context.AvailableChargeLinkReceiptData.AddRangeAsync(availableChargeLinkReceiptData);
+            await _context.SaveChangesAsync();
+        }
+
+        public Task<List<AvailableChargeLinkReceiptData>> GetAvailableChargeLinkReceiptDataAsync(IEnumerable<Guid> dataReferenceId)
+        {
+            var queryable = _context.AvailableChargeLinkReceiptData.Where(x => dataReferenceId.Contains(x.AvailableDataReferenceId));
+            return queryable
+                .OrderBy(x => x.RequestDateTime)
+                .ToListAsync();
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/AvailableChargeLinkReceiptDataRepository.cs:45]:
```suggestion
            return _context
                .AvailableChargeLinkReceiptData
                .Where(x => dataReferenceId.Contains(x.AvailableDataReferenceId))
                .OrderBy(x => x.RequestDateTime)
                .ToListAsync();
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/RepositoryAutoMoqDataFixer.cs [hunk 9]
```diff
@@ -37,9 +41,28 @@ public static AvailableChargeData FixAvailableChargeData([NotNull] AvailableChar
                 availableChargeData.TaxIndicator,
                 availableChargeData.TransparentInvoicing,
                 availableChargeData.Resolution,
-                availableChargeData.Points.ToList(),
-                availableChargeData.RequestTime,
-                availableChargeData.AvailableDataReferenceId);
+                availableChargeData.Points.ToList());
+        }
+
+        public static AvailableChargeLinkReceiptData FixAvailableChargeLinkReceiptData(
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/RepositoryAutoMoqDataFixer.cs:47]: The name seems to suggest that this method modifies the input argument. So I suggest either doing that or changing the name.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/AvailableChargeLinkReceiptDataRepositoryTests.cs [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+        [Theory]
+        [InlineAutoMoqData]
+        public async Task StoreAsync_StoresAvailableChargeLinkReceiptData([NotNull] List<AvailableChargeLinkReceiptData> expectedList)
+        {
+            // Arrange
+            expectedList = RepositoryAutoMoqDataFixer.FixAvailableChargeLinkReceiptDataList(expectedList);
+
+            await using var chargesDatabaseWriteContext = _databaseManager.CreateDbContext();
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/AvailableChargeLinkReceiptDataRepositoryTests.cs:47]: Did we ever write a test that verifies that each call to the database manager to create a db context returns a new context each time? (To ensure that we are not e.g. just asserting on in-memory state of the write context)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/AvailableChargeLinkReceiptDataRepositoryTests.cs [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            await sut.StoreAsync(expectedList).ConfigureAwait(false);
+
+            // Assert
+            await using var chargesDatabaseReadContext = _databaseManager.CreateDbContext();
+
+            foreach (var expected in expectedList)
+            {
+                var actual =
+                    chargesDatabaseReadContext.AvailableChargeLinkReceiptData
+                        .Single(x => x.AvailableDataReferenceId == expected.AvailableDataReferenceId);
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/AvailableChargeLinkReceiptDataRepositoryTests.cs:61]:
```suggestion
                var actual = chargesDatabaseReadContext
                    .AvailableChargeLinkReceiptData
                    .Single(x => x.AvailableDataReferenceId == expected.AvailableDataReferenceId);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/AvailableChargeLinkReceiptDataRepositoryTests.cs [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+                actual.RecipientId.Should().Be(expected.RecipientId);
+                actual.RecipientRole.Should().Be(expected.RecipientRole);
+                actual.BusinessReasonCode.Should().Be(expected.BusinessReasonCode);
+                actual.RequestDateTime.Should().Be(expected.RequestDateTime);
+                actual.AvailableDataReferenceId.Should().Be(expected.AvailableDataReferenceId);
+                actual.ReceiptStatus.Should().Be(expected.ReceiptStatus);
+                actual.OriginalOperationId.Should().Be(expected.OriginalOperationId);
+                actual.MeteringPointId.Should().Be(expected.MeteringPointId);
+                actual.ReasonCodes.Should().BeEquivalentTo(expected.ReasonCodes);
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/AvailableChargeLinkReceiptDataRepositoryTests.cs:70]: Can't this be optimized using e.g. https://fluentassertions.com/objectgraphs/?
Would be nice if it becomes less prone to forgetting to add new props.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargesDatabaseContext.cs (hunk 6 above)
Comment (bemwen) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargesDatabaseContext.cs:56]: Yes it does. Both of them.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/AvailableChargeData.cs (hunk 1 above)
Comment (bemwen) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/AvailableChargeData.cs]: Fixed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/AvailableChargeDataFactory.cs (hunk 2 above)
Comment (bemwen) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/AvailableChargeDataFactory.cs]: Fixed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeLinkReceiptData/IAvailableChargeLinkReceiptDataRepository.cs (hunk 3 above)
Comment (bemwen) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeLinkReceiptData/IAvailableChargeLinkReceiptDataRepository.cs]: Changed, in all three available data repositories

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeLinksData/AvailableChargeLinksDataFactory.cs (hunk 4 above)
Comment (bemwen) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeLinksData/AvailableChargeLinksDataFactory.cs]: Fixed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/AvailableChargeLinkReceiptDataRepositoryTests.cs (hunk 10 above)
Comment (bemwen) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/AvailableChargeLinkReceiptDataRepositoryTests.cs:47]: I would consider this out of scope of this PR and defer it to someone with better understanding of EF.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/EntityConfigurations/AvailableChargeLinkReceiptDataConfiguration.cs (hunk 7 above)
Comment (bemwen) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/EntityConfigurations/AvailableChargeLinkReceiptDataConfiguration.cs:37]: Hmm, spending to much time on this now, skipping this for now and returning in a later PR.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/RepositoryAutoMoqDataFixer.cs (hunk 9 above)
Comment (bemwen) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/Repositories/RepositoryAutoMoqDataFixer.cs]: Naming changed to better fit use

**Review by BjarkeMeier — APPROVED** [review]
LGTM

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/AvailableChargeData.cs [hunk 13]
```diff
@@ -14,18 +14,21 @@
...
 using GreenEnergyHub.Charges.Domain.Charges;
 using GreenEnergyHub.Charges.Domain.MarketParticipants;
 using NodaTime;
 
 namespace GreenEnergyHub.Charges.Domain.AvailableChargeData
 {
-    public class AvailableChargeData
+    public class AvailableChargeData : AvailableDataBase
     {
         public AvailableChargeData(
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/AvailableChargeData.cs:26]: Should we check properties are not null, which are not allowed to be null in the database?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/IAvailableChargeDataRepository.cs [hunk 14]
```diff
@@ -22,6 +22,6 @@ public interface IAvailableChargeDataRepository
     {
         Task StoreAsync(AvailableChargeData availableChargeData);
 
-        Task<List<AvailableChargeData>> GetAvailableChargeDataAsync(IEnumerable<Guid> dataReferenceId);
+        Task<List<AvailableChargeData>> GetAsync(IEnumerable<Guid> dataReferenceId);
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/IAvailableChargeDataRepository.cs:25]: ids / dataReferenceIds

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeLinksData/IAvailableChargeLinksDataRepository.cs [hunk 15]
```diff
@@ -22,6 +22,6 @@ public interface IAvailableChargeLinksDataRepository
     {
         Task StoreAsync(List<AvailableChargeLinksData> availableChargeLinksData);
 
-        Task<List<AvailableChargeLinksData>> GetAvailableChargeLinksDataAsync(IEnumerable<Guid> dataReferenceId);
+        Task<List<AvailableChargeLinksData>> GetAsync(IEnumerable<Guid> dataReferenceId);
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeLinksData/IAvailableChargeLinksDataRepository.cs:25]: dataReferenceId -> ids / dataReferenceIds

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/AvailableChargeDataRepository.cs [hunk 16]
```diff
@@ -37,12 +37,12 @@ public async Task StoreAsync(AvailableChargeData availableChargeData)
             await _context.SaveChangesAsync();
         }
 
-        public Task<List<AvailableChargeData>> GetAvailableChargeDataAsync(IEnumerable<Guid> dataReferenceId)
+        public Task<List<AvailableChargeData>> GetAsync(IEnumerable<Guid> dataReferenceId)
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/AvailableChargeDataRepository.cs:40]: dataReferenceId -> ids / dataReferenceIds

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/AvailableChargeLinkReceiptDataRepository.cs [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+            _context = context;
+        }
+
+        public async Task StoreAsync(List<AvailableChargeLinkReceiptData> availableChargeLinkReceiptData)
+        {
+            await _context.AvailableChargeLinkReceiptData.AddRangeAsync(availableChargeLinkReceiptData);
+            await _context.SaveChangesAsync();
+        }
+
+        public Task<List<AvailableChargeLinkReceiptData>> GetAsync(IEnumerable<Guid> dataReferenceId)
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/AvailableChargeLinkReceiptDataRepository.cs:40]: dataReferenceId -> ids / dataReferenceIds

**Review by lasrinnil — APPROVED** [review]

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/IAvailableChargeDataRepository.cs (hunk 14 above)
Comment (bemwen) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/AvailableChargeData/IAvailableChargeDataRepository.cs:25]: Adressing these in next PR

**Pull request merged**